### ISSUE TITLE
Distinguishing zero results because of no results, or because of server error

### ISSUE THREAD

**Issue by dstansby** [sunpy/sunpy#5661]
This weekend the solar orbiter archive was down, which prompted this bug report for a not very nice error being raised by `sunpy-soar`: https://github.com/dstansby/sunpy-soar/issues/11. I think the solution to that issue is to return zero results from the client, so other clients have a chance to return results (e.g. for a time only search). It would be helpful to have a way of informing the user of the difference between:

- Returned zero results because the server didn't have any results
- Returned zero results because of a server error

**Comment by Cadair** [comment]
We could do better, but what some clients do is to raise warnings and then return 0 results.

**Comment by dstansby** [comment]
I am 👍 to resolving this by recommending people writing clients raise a warning if the server they are querying is down, and by putting such warnings in the sunpy core clients.

**Comment by nabobalis** [comment]
> @nabobalis , i would like to give this issue a try

Sure but I would suggest tackling a new issue one at a time especially if you are working on other issues.

**Comment by exitflynn** [comment]
if this issue isn't being worked on, i'd like to give this a go. I'm not sure if we can assume a specific error type for every source that could go down in the future? I suppose one way to implement this would be to just verify if the source has returned any (or zero) results and raise an error if not.

**Comment by ryuusama09** [comment]
@nabobalis , I see an open pr here , but it seems too old so I hope i can do something here ?

**Comment by nabobalis** [comment]
> @nabobalis , I see an open pr here , but it seems too old so I hope i can do something here ?

I don't see an open PR?

**Comment by ryuusama09** [comment]
oops ! . I think I commented for some other issue. Anyways thank you , going ahead with this !

**Comment by ryuusama09** [comment]
@nabobalis , dont you think this issue is similar to #7279. While we try to raise warnings in the ```search()``` method in this issue , we try to do the something similar ```fetch()``` . (the similarity is that ```fetch``` fails to download and ```search``` fails to return results because of server issue. the only different part is handling the ```0 result``` in search ).

**Comment by nabobalis** [comment]
I would say it's similar but with different intentions and behaviors. 
But I don't see the connection and making a decision on both?

**Comment by ryuusama09** [comment]
> But I don't see the connection and making a decision on both?

There was a huge discussion on raising warnings/errors/interrupts . I thought similar could be applied here . ( a debate on whether the user should do the error handling or provide it ). If this is not the case , then can I go ahead with implementing error handling block in ```search()``` ?

**Comment by nabobalis** [comment]
> There was a huge discussion on raising warnings/errors/interrupts . I thought similar could be applied here . ( a debate on whether the user should do the error handling or provide it ). If this is not the case , then can I go ahead with implementing error handling block in `search()` ?

But as far as I can tell, no error will be raised here. We will try to issue a warning depending on if a search result is blank due to a server error or if there was 0 results due to the search.

**Comment by ryuusama09** [comment]
yep so if ```len(results) == 0``` we raise a 0 matched results warning . Incase it is a server error , how does ```search``` method handle it currently ? . (I was thinking of raising a warning when there is a timeout ).

**Comment by nabobalis** [comment]
> right now , if I am making an invalid query , the client is running indefinitely . 

Really? it doesn't return?

>Using this we can say that -> if `result._numfile == 0` then its a server error -> if `len(results) == 0 ` then 0 matching files Or they could be meaning the same thing .

I would be surprised if `result._numfile == 0` would be a server error. 
You will have to investigate further.

**Comment by nabobalis** [comment]
> > I would be surprised if `result._numfile == 0` would be a server error.
> > You will have to investigate further.
> 
> Yes, Thought as much . I think essentially both meant the same thing. In one case server error could be returning an error message .. currently I didn't find any error attribute in the `unifiedresponse` object which is being returned. In another case server error could mean no response at all . I don't know how that is being handled currently .

In that case, it might be better to break the code to mimic a server error and see what happens.

> Nope , It ran for like > minute , which ultimately i had to stop the execution.

What was the query you ran?

**Comment by nabobalis** [comment]
Does the scarper not raise an error if something fails?

**Comment by nabobalis** [comment]
Hard to say without the context of the code around it. If there is a 404, it might not be a server error, the file just might be missing. So I don't see why we would raise that as an error.

**Comment by nabobalis** [comment]
I think I have myself confused. 

So if we want to achieve a different warnings based on: 

Returned zero results because the server didn't have any results
Returned zero results because of a server error

We first need to work out what the default behaviors are for each client if the server returns an error vs if the query is valid but there is no data. Is this for example only a scraper client problem or do the VSO or JSOC clients also need this? 

From there, we would need to work out the behavior we want and then after that I don't have anything concrete.

**Comment by nabobalis** [comment]
> > Returned zero results because the server didn't have any results
> 
> This is simple to mimic , for example picking up a time range for year 2025. It returns an empty `meta_list`.
> 
> > Returned zero results because of a server error
> 
> This is difficult to mimic . Even If we break the code , there is a good chance that server might handle the exception and provide some response. Typically I am struggling to get the error message returned from the server . I think we should be returning the error code from scraper , else it is difficult to understand if it is servers issue.

Ok, so what will you do now?

**Comment by nabobalis** [comment]
> I feel , server error could only be found be actually returning the server error response. Right now if a directory gives an error it is queued back for request. Is there a mechanism where there is a limit on requeueing ?

I don't recall, can you check the scraper code?

**Comment by nabobalis** [comment]
> I didn't find one I think . In that case , we can limit the requeueing , and once the limit is reached return error . One thing to look is that even if one directory fails , should we still return the partial data or right away throw `0` data ? Cases like these would be needed to be handled.

If a client finds partial data, we should return that but maybe with a warning saying it would be incomplete.

**Comment by nabobalis** [comment]
> Alright . Then should I proceed with what we discussed above and raise a draft pr ?

Sure, I still have some doubts but when I see something, it should be easier to work them out.

**Comment by nabobalis** [comment]
That only works on Python 3.10 if I recall which the main branch of sunpy is now using as the lowest version.

**Comment by nabobalis** [comment]
To python 3.10?

**Comment by nabobalis** [comment]
I am using python 3.12 and do not have this issue.

Are you sure you are using Python 3.10?

**Comment by samaloney** [comment]
I think this is partially addressed by #8009 so if the clients return errors it won't stop the search in other client and they are displayed to the user too.
